pages/AddressBook.qml

```diff
@@ -58,18 +62,73 @@ Rectangle {
                 }
             }
 
-            LineEditMulti {
-                Layout.fillWidth: true;
+            MoneroComponents.LineEditMulti {
                 id: addressLine
                 labelText: qsTr("Address") + translationManager.emptyString
-                error: true;
-                placeholderText: qsTr("4.. / 8..") + translationManager.emptyString
+                placeholderText: "4.. / 8.. / OpenAlias"
                 wrapMode: Text.WrapAnywhere
                 addressValidation: true
+                pasteButton: true
+                onPaste: function(clipboardText) {
+                    const parsed = walletManager.parse_uri_to_object(clipboardText);
+                    if (!parsed.error) {
+                        addressLine.text = parsed.address;
+                        setPaymentId(parsed.payment_id);
+                        setDescription(parsed.tx_description);
+                    } else {
+                        addressLine.text = clipboardText;
+                    }
+                }
             }
         }
 
-        LineEditMulti {
+        MoneroComponents.StandardButton {
+            id: resolveButton
+            text: qsTr("Resolve") + translationManager.emptyString
+            visible: TxUtils.isValidOpenAliasAddress(addressLine.text)
+            enabled : visible
+            onClicked: {
+                var result = walletManager.resolveOpenAlias(addressLine.text)
```

Comment:
Would it be possible remove the code duplication here?

Reply:
Do you mean resolveOpenAlias vs isValidOpenAliasAddress? Not really as they do different things.

Reply:
```
                var result = walletManager.resolveOpenAlias(addressLine.text)
	                if (result) {
	                    var parts = result.split("|")
	                    if (parts.length == 2) {
	                        var address_ok = walletManager.addressValid(parts[1], appWindow.persistentSettings.nettype)
	                        if (parts[0] === "true") {
	                            if (address_ok) {
	                                // prepend openalias to description
	                                descriptionLine.text = descriptionLine.text ? addressLine.text + " " + descriptionLine.text : addressLine.text
	                                addressLine.text = parts[1]
	                                addressLine.cursorPosition = 0
	                            }
	                            else
	                                oa_message(qsTr("No valid address found at this OpenAlias address"))
	                        }
	                        else if (parts[0] === "false") {
	                              if (address_ok) {
	                                  addressLine.text = parts[1]
	                                  addressLine.cursorPosition = 0
	                                  oa_message(qsTr("Address found, but the DNSSEC signatures could not be verified, so this address may be spoofed"))
	                              }
	                              else
	                              {
	                                  oa_message(qsTr("No valid address found at this OpenAlias address, but the DNSSEC signatures could not be verified, so this may be spoofed"))
	                              }
	                        }
	                        else {
	                            oa_message(qsTr("Internal error"))
	                        }
	                    }
	                    else {
	                        oa_message(qsTr("Internal error"))
	                    }
	                }
	                else {
	                    oa_message(qsTr("No address found"))
	                }
```

This whole code block is in `pages/Transferl.qml` and `pages/AddressBook.qml` now.

Reply:
Not worth it in this PR. Transfer page is changing quite a bit still. Better after. In a future PR, this could be reorganized and refactored into a component. Also this definitely could be refactored to be more readable.